File: 6. Tree/Binary Search Tree/binarySearchTree.js
const prompt = require('prompt-sync')({ sigint: true });

class Node {
    constructor(val) {
        this.data = val;
        this.right = null;
        this.left = null;
    }
}

let root = null;

function getInOrderSuccessor(node) {
    while (node != null && node.left != null)
        node = node.left;
    return node;
}

function createNode(data) {
    let temp = new Node(data);
    return temp;
}

function insert() { // Insert
    let value = Number.parseInt(prompt("Enter Value:- "));

    let newNode = createNode(value);

    if (root == null) {
        root = newNode;
        console.log("Insertion Successful. Value: ", value);
        return;
    }

    let temp = root;

    while (true) {
        if (temp.data == value) {
            System.out.println("Value already exists in the tree.\n");
            return;
        } else if (temp.data > value) {
            if (temp.left == null) {
                temp.left = newNode;
                break;
            }
            temp = temp.left;
        } else if (temp.data < value) {
            if (temp.right == null) {
                temp.right = newNode;
                break;
            }
            temp = temp.right;
        }
    }

    console.log("Insertion Successful. Value: %d\n", value);
}

function deleteNode(root, key, isDeleted) { // Remove 
    if (root == null)
        return null;

    if (key < root.data) {
        root.left = deleteNode(root.left, key, isDeleted);
    } else if (key > root.data) {
        root.right = deleteNode(root.right, key, isDeleted);
    } else {
        isDeleted.bool = true;
        // Case 1: No child or Case 2: One child
        if (root.left == null) {
            let temp = root.right;
            return temp;
        } else if (root.right == null) {
            let temp = root.left;
            return temp;
        }

        // Case 3: Two children
        let succ = getInOrderSuccessor(root.right);
        root.data = succ.data;
        root.right = deleteNode(root.right, succ.data, isDeleted);
    }

    return root;
}

function search() { // Search 
    if (root == null) {
        console.log("\n~~Tree is Empty~~");
        return;
    }

    let val = prompt("Enter Value:- ");

    temp = root;

    while (true) {
        if (temp.data == val) {
            console.log(`Element Found: ${temp.data}\n`);
            return;
        } else if (temp.data > val) {
            temp = temp.left;
        } else if (temp.data < val) {
            temp = temp.right;
        }

        if (temp == null) {
            console.log("Element not found!\n");
            return;
        }
    }
}

function inOrderTraversal(node) {
    if (node != null) {
        inOrderTraversal(node.left);
        console.log(`${node.data} `);
        inOrderTraversal(node.right);
    }
    return;
}

function preOrderTraversal(node) {
    if (node != null) {
        console.log(`${node.data} `);
        preOrderTraversal(node.left);
        preOrderTraversal(node.right);
    }
    return;
}

function postOrderTraversal(node) {
    if (node != null) {
        postOrderTraversal(node.left);
        postOrderTraversal(node.right);
        console.log(`${node.data} `);
    }
    return;
}

function levelOrderTraversal() {
    let queue = [];
    let front = 0, rear = 0;

    queue[rear++] = root;

    let temp;

    while (front < rear) {
        temp = queue[front++];

        console.log("%d ", temp.data);

        if (temp.left) queue[rear++] = temp.left;
        if (temp.right) queue[rear++] = temp.right;
    }

    return;
}

(() => {
    let choice, choice2;
    console.log("Hello! Here You Can Perform Following Binary Search Tree Operation!");
    while (true) {
        console.log("\n1. Show Data\n2. Insert\n3. Delete\n4. Search\n5. Exit");
        choice = Number.parseInt(prompt("Enter Your Choice:- "));
        switch (choice) {
            case 1:
                if (root == null) {
                    console.log("\n~~Tree is Empty~~");
                    break;
                }

                console.log("\n1. In-Order Traversal(DFS)\n2. Pre-Order Traversal(DFS)\n3. Post-Order Traversal(DFS)\n4. Level-Order Traversal(BFS)");
                choice2 = Number.parseInt(prompt("Enter Your Choice:- "));
                switch (choice2) {
                    case 1:
                        console.log("\nIn-Order Traversal(DFS): ");
                        inOrderTraversal(root);
                        break;
                    case 2:
                        console.log("\nPre-Order Traversal(DFS): ");
                        preOrderTraversal(root);
                        break;
                    case 3:
                        console.log("\nPost-Order Traversal(DFS): ");
                        postOrderTraversal(root);
                        break;
                    case 4:
                        console.log("\nLevel-Order Traversal(BFS): ");
                        levelOrderTraversal();
                        break;
                    default:
                        console.log(`\n~~ Please Enter a Vaild Input! ~~\n`);
                        break;
                }
                break;
            case 2:
                insert();
                break;
            case 3:
                if (root == null) {
                    console.log("\n~~Tree is Empty~~\n");
                    break;
                }

                let val = prompt("Enter Value:- ");

                /**
                 * Used Object, Because as we passes it into Func the reference is Pass. Help to easily track deletion
                 */
                let isDeleted = { bool: false };
                
                root = deleteNode(root, val, isDeleted);
                
                if (!isDeleted.bool) {
                    console.log("Element not found!\n");
                } else {
                    console.log("Element Deleted: %d\n", val);
                }

                break;
            case 4:
                search();
                break;
            case 5:
                console.log(`\n~~ Thanks For Using! Have a Great Day! ~~\n`);
                return;
            default:
                console.log(`\n~~ Please Enter a Vaild Input! ~~\n`);
                break;
        }
    }
    return;
})();

File: 6. Tree/Binary Search Tree/binarySearchTree.py
class Node:
    def __init__(self, data):
        self.data = data;
        self.right = None;
        self.left = None;

root = None;
isDeleted = False;

def getInOrderSuccessor(node):
    while (node != None and node.left != None):
            node = node.left;
    return node;

def createNode(data):
    temp = Node(data);
    return temp;

def insert():
    global root;

    val = int(input("Enter Value:- "));

    newNode = createNode(val);

    if (root == None):
        root = newNode;
        print("\nInsertion Successfull!");
        return;

    temp = root;

    while True:
        if (temp.data == val):
                print("Value already exists in the tree.\n");
                return;
        elif (temp.data > val):
            if (temp.left == None):
                temp.left = newNode;
                break;
            temp = temp.left;
        elif (temp.data < val):
            if (temp.right == None):
                    temp.right = newNode;
                    break;
            temp = temp.right;

    print("\nInsertion Successfull!");

def deleteNode(root, key):
    global isDeleted;
    if (root == None):
        return None;

    if (key < root.data):
        root.left = deleteNode(root.left, key);
    elif (key > root.data):
        root.right = deleteNode(root.right, key);
    else:
        isDeleted = True;

        # Case 1: No child or Case 2: One child
        if (root.left == None):
            temp = root.right;
            return temp;
        elif (root.right == None):
            temp = root.left;
            return temp;

        # Case 3: Two children
        succ = getInOrderSuccessor(root.right);
        root.data = succ.data;
        root.right = deleteNode(root.right, succ.data);

    return root;

def search():
    global root;

    if (root == None):
        print("\n~~Tree is Empty~~");
        return;

    val = int(input("Enter Value:- "));

    temp = root;

    while True:
        if (temp.data == val):
            print("Element Found: ", temp.data);
            return;
        elif (temp.data > val):
            temp = temp.left;
        elif (temp.data < val):
            temp = temp.right;

        if (temp == None):
            print("Element Not Found!");
            return;

def inOrderTraversal(node):
    if (node != None):
        inOrderTraversal(node.left);
        print(node.data, " ", end="");
        inOrderTraversal(node.right);
    return;

def preOrderTraversal(node):
    if (node != None):
        print(node.data, " ", end="");
        preOrderTraversal(node.left);
        preOrderTraversal(node.right);
    return;

def postOrderTraversal(node):
    if (node != None):
        postOrderTraversal(node.left);
        postOrderTraversal(node.right);
        print(node.data, " ", end="");
    return;

def levelOrderTraversal():
    queue = [];
    front = 0;
    rear = 0;

    queue.insert(rear, root);
    rear += 1;

    temp = None;

    while (front < rear):
        temp = queue[front];
        front += 1;

        print(temp.data, " ", end="");

        if (temp.left):
            queue.insert(rear, temp.left);
            rear += 1;
        if (temp.right):
            queue.insert(rear, temp.right);
            rear += 1;

    return;

print("Hello! Here You Can Perform Following Binary Tree Operation!");
while (True):
    print("\n1. Show Data\n2. Insert\n3. Delete\n4. Search\n5. Exit");
    choice = int(input("Enter Your Choice:- "));
    if(choice == 1):
        if (root == None):
            print("\n~~Tree is Empty~~");
            continue;
        
        print("\n1. In-Order Traversal(DFS)\n2. Pre-Order Traversal(DFS)\n3. Post-Order Traversal(DFS)\n4. Level-Order Traversal(BFS)");
        choice2 = int(input("Enter Your Choice:- "));
        if(choice2 == 1):
            print();
            inOrderTraversal(root);
            print();
            continue;
        elif(choice2 == 2):
            print();
            preOrderTraversal(root);
            print();
            continue;
        elif(choice2 == 3):
            print();
            postOrderTraversal(root);
            print();
            continue;
        elif(choice2 == 4):
            print();
            levelOrderTraversal();
            print();
            continue;
        else:
            print("\n~~ Please Enter a Vaild Input! ~~\n");
            continue;
    elif(choice == 2):
        insert();
        continue;
    elif(choice == 3):
        if (root == None):
            print("\n~~Tree is Empty~~");
            continue;

        val = int(input("Enter Value:- "));

        isDeleted = False;

        root = deleteNode(root, val);

        if not isDeleted:
            print("Element not found!\n");
        else:
            print("Element Deleted: %d\n", val);

        continue;
    elif(choice == 4):
        search();
        continue;
    elif(choice == 5):
        print("\n~~ Thanks For Using! Have a Great Day! ~~\n");
        break;
    else:
        print("\n~~ Please Enter a Vaild Input! ~~\n");
        continue;